File: js/GameForm.js
/**
 * GameForm.js — Main game controller (port of Forms/GameForm.cs)
 * 
 * Replaces WinForms UI manipulation with DOM operations.
 * This is the central orchestrator: movement, actions, combat delegation, UI updates.
 */
class GameForm {
    constructor(player, gameState = null) {
        this.currentPlayer = player;
        this.gameState = gameState || new GameState();
        this.maze = new Maze();
        this.playerX = this.gameState.playerX;
        this.playerY = this.gameState.playerY;
        this.currentCombat = null;
        this.isGameOver = false;
        this.selectedAbilityIndex = -1;

        // Cache DOM references
        this.els = {
            gameLog:        document.getElementById('txtGameLog'),
            playerName:     document.getElementById('lblPlayerName'),
            playerHP:       document.getElementById('pbPlayerHealth'),
            playerMaxHP:    document.getElementById('lblPlayerMaxHP'),
            playerStr:      document.getElementById('lblPlayerStrength'),
            playerDex:      document.getElementById('lblPlayerDexterity'),
            playerDef:      document.getElementById('lblPlayerDefense'),
            playerWis:      document.getElementById('lblPlayerWisdom'),
            playerAbilities:document.getElementById('lbPlayerAbilities'),
            inventory:      document.getElementById('lbInventory'),
            actionsTaken:   document.getElementById('lbActionsTaken'),
            enemyPanel:     document.getElementById('pnlEnemyStats'),
            enemyName:      document.getElementById('lblEnemyName'),
            enemyHP:        document.getElementById('pbEnemyHealth'),
            enemyMaxHP:     document.getElementById('lblEnemyMaxHP'),
            enemyStr:       document.getElementById('lblEnemyStrength'),
            enemyDex:       document.getElementById('lblEnemyDexterity'),
            enemyDef:       document.getElementById('lblEnemyDefense'),
            enemyWis:       document.getElementById('lblEnemyWisdom'),
            enemyAbilities: document.getElementById('lbEnemyAbilities'),
            pnlMovement:    document.getElementById('pnlMovement'),
            pnlCombat:      document.getElementById('pnlCombat'),
            btnUp:          document.getElementById('btnUp'),
            btnDown:        document.getElementById('btnDown'),
            btnLeft:        document.getElementById('btnLeft'),
            btnRight:       document.getElementById('btnRight'),
            btnAction:      document.getElementById('btnAction'),
            btnFight:       document.getElementById('btnFight'),
            btnUseAbility:  document.getElementById('btnUseAbility'),
            txtX:           document.getElementById('txtX'),
            txtY:           document.getElementById('txtY')
        };

        this.bindEvents();
        this.showMovementUI();
        this.updatePlayerUI();
        this.refreshInventoryList();
        this.refreshActionsTakenList();
        this.enterRoom();
    }

    // ============ Event Binding ============

    bindEvents() {
        this.els.btnUp.addEventListener('click', () => this.movePlayer(0, 1));
        this.els.btnDown.addEventListener('click', () => this.movePlayer(0, -1));
        this.els.btnLeft.addEventListener('click', () => this.movePlayer(-1, 0));
        this.els.btnRight.addEventListener('click', () => this.movePlayer(1, 0));

        this.els.btnFight.addEventListener('click', () => {
            if (this.currentCombat) this.currentCombat.playerAttack();
        });

        this.els.btnUseAbility.addEventListener('click', () => {
            if (this.currentCombat && this.selectedAbilityIndex >= 0) {
                const ability = this.currentPlayer.abilities[this.selectedAbilityIndex];
                if (ability) this.currentCombat.playerUseAbility(ability);
            } else {
                this.appendLog('Select an ability first.');
            }
        });

        this.els.btnAction.addEventListener('click', () => this.handleAction());

        document.getElementById('btnSaveQuit').addEventListener('click', () => {
            this.saveGame();
            this.appendLog('Game saved.');
            // Switch to start screen
            switchScreen('screen-start');
            checkResumeButton();
        });

        document.getElementById('btnStartOver').addEventListener('click', () => {
            if (confirm('Are you sure you want to start over? This will erase all game progress.')) {
                this.resetGame();
            }
        });

        // Modal play-again button
        document.getElementById('modalBtnYes').addEventListener('click', () => {
            bootstrap.Modal.getInstance(document.getElementById('gameModal')).hide();
            this.resetGame();
        });

        document.getElementById('modalBtnNo').addEventListener('click', () => {
            switchScreen('screen-start');
            checkResumeButton();
        });
    }

    // ============ Core Logic ============

    startCombat(enemies) {
        this.currentCombat = new CombatEngine(this, this.currentPlayer, enemies);
        this.currentCombat.startCombat();
        this.showCombatUI();
    }

    enterRoom() {
        this.updateMovementButtons();
        this.els.txtX.textContent = this.playerX;
        this.els.txtY.textContent = this.playerY;

        const room = this.maze.getRoom(this.playerX, this.playerY);
        if (room) {
            this.appendLog(`\n=== ${room.name} ===`, 'heading');
            if (room.description) this.appendLog(room.description);
            if (room.roomEnterLogic) room.roomEnterLogic(this);
            if (room.roomUILogic) room.roomUILogic(this);
        }
    }

    loadRoom() {
        this.updateMovementButtons();
        this.els.txtX.textContent = this.playerX;
        this.els.txtY.textContent = this.playerY;

        const room = this.maze.getRoom(this.playerX, this.playerY);
        if (room && room.roomUILogic) room.roomUILogic(this);
    }

    teleportTo(x, y) {
        this.playerX = x;
        this.playerY = y;
        this.gameState.playerX = x;
        this.gameState.playerY = y;
        this.enterRoom();
    }

    // ============ Movement ============

    movePlayer(dx, dy) {
        const newX = this.playerX + dx;
        const newY = this.playerY + dy;

        const room = this.maze.getRoom(newX, newY);
        if (!room) {
            this.appendLog("You can't go that way.");
            return;
        }

        this.playerX = newX;
        this.playerY = newY;

        this.gameState.playerX = this.playerX;
        this.gameState.playerY = this.playerY;
        this.saveGame();

        this.enterRoom();
    }

    setMovementEnabled(enabled) {
        if (enabled) {
            this.updateMovementButtons();
        } else {
            this.els.btnUp.disabled = true;
            this.els.btnDown.disabled = true;
            this.els.btnLeft.disabled = true;
            this.els.btnRight.disabled = true;
        }
    }

    updateMovementButtons() {
        this.els.btnUp.disabled    = !this.maze.getRoom(this.playerX, this.playerY + 1);
        this.els.btnDown.disabled  = !this.maze.getRoom(this.playerX, this.playerY - 1);
        this.els.btnRight.disabled = !this.maze.getRoom(this.playerX + 1, this.playerY);
        this.els.btnLeft.disabled  = !this.maze.getRoom(this.playerX - 1, this.playerY);
    }

    lockDirection(dir) {
        switch (dir) {
            case 'Left':  this.els.btnLeft.disabled = true; break;
            case 'Right': this.els.btnRight.disabled = true; break;
            case 'Up':    this.els.btnUp.disabled = true; break;
            case 'Down':  this.els.btnDown.disabled = true; break;
        }
    }

    unlockDirection(dir) {
        switch (dir) {
            case 'Left':  this.els.btnLeft.disabled = false; break;
            case 'Right': this.els.btnRight.disabled = false; break;
            case 'Up':    this.els.btnUp.disabled = false; break;
            case 'Down':  this.els.btnDown.disabled = false; break;
        }
    }

    // ============ Actions ============

    handleAction() {
        const actionText = this.els.btnAction.textContent;

        switch (actionText) {
            case 'Route Power to Sensors':
                if (!this.gameState.actionsTaken.includes('RoutedPower'))
                    this.gameState.actionsTaken.push('RoutedPower');
                this.appendLog('You reroute auxiliary power to the sensor array. The console lights up.');
                break;

            case 'Take Phaser':
                if (!this.gameState.inventory.includes('Phaser'))
                    this.gameState.inventory.push('Phaser');
                this.appendLog('You take the Type-2 Phaser.');
                break;

            case 'Cut Blast Door':
                if (!this.gameState.actionsTaken.includes('CutDoor'))
                    this.gameState.actionsTaken.push('CutDoor');
                this.appendLog('You set the Phaser to maximum and slice through the blast door locks. 💥');
                this.appendLog('The door slides open with a groan.');
                break;

            case 'Engage Override':
                if (!this.gameState.actionsTaken.includes('Engaged override')) {
                    this.gameState.actionsTaken.push('Engaged override');
                    this.appendLog('You engage the manual override. The forcefields deactivate.');
                } else {
                    this.appendLog('Override already engaged.');
                }
                break;

            case 'Take Keycard':
                if (!this.gameState.inventory.includes('Keycard'))
                    this.gameState.inventory.push('Keycard');
                this.appendLog('You pick up the Security Keycard.');
                break;

            case 'Take Tricorder':
                if (!this.gameState.inventory.includes('Tricorder'))
                    this.gameState.inventory.push('Tricorder');
                this.appendLog('You pick up the standard issue Tricorder.');
                break;

            case 'Access Locker':
                this.gameWin();
                return; // Don't continue to save/refresh

            case 'Scan Entity':
                this.appendLog('You scan the entity. It appears to be a silicon-based lifeform.');
                break;

            case 'Take Combadge':
                if (!this.gameState.inventory.includes('Combadge')) {
                    this.gameState.inventory.push('Combadge');
                    this.appendLog('You take the Combadge from the dead Redshirt.');
                    this.appendLog("You feel a strange connection to the ship's engineer...");
                    this.currentPlayer.abilities.push(new CallScottyAbility());
                    this.appendLog('Ability Acquired: Call Scotty! 🖖');
                    this.hideActionButton();
                    this.updatePlayerUI();
                }
                break;

            default:
                this.appendLog(`Action '${actionText}' not handled.`);
                break;
        }

        this.saveGame();
        this.loadRoom();
        this.refreshInventoryList();
        this.refreshActionsTakenList();
    }

    // ============ UI Management ============

    updatePlayerUI() {
        if (!this.currentPlayer) return;
        const p = this.currentPlayer;

        const pct = Math.max(0, Math.min(100, (p.currentHP / Math.max(1, p.maxHP)) * 100));
        this.els.playerHP.style.width = pct + '%';
        this.els.playerName.textContent = p.name;
        this.els.playerMaxHP.textContent = `${p.currentHP}/${p.maxHP}`;
        this.els.playerStr.textContent = `STR: ${p.strength}`;
        this.els.playerDex.textContent = `DEX: ${p.dexterity}`;
        this.els.playerDef.textContent = `DEF: ${p.defense}`;
        this.els.playerWis.textContent = `WIS: ${p.wisdom}`;

        this.updatePlayerAbilitiesList();
    }

    updateEnemyUI(enemy) {
        if (!enemy) {
            this.showMovementUI();
            return;
        }

        this.showCombatUI();
        const pct = Math.max(0, Math.min(100, (enemy.currentHP / Math.max(1, enemy.maxHP)) * 100));
        this.els.enemyHP.style.width = pct + '%';
        this.els.enemyName.textContent = enemy.name;
        this.els.enemyMaxHP.textContent = `${enemy.currentHP}/${enemy.maxHP}`;
        this.els.enemyStr.textContent = `STR: ${enemy.strength}`;
        this.els.enemyDex.textContent = `DEX: ${enemy.dexterity}`;
        this.els.enemyDef.textContent = `DEF: ${enemy.defense}`;
        this.els.enemyWis.textContent = `WIS: ${enemy.wisdom}`;

        // Enemy abilities
        this.els.enemyAbilities.innerHTML = '';
        if (enemy.abilities) {
            enemy.abilities.forEach(a => {
                const li = document.createElement('li');
                li.textContent = a.name;
                this.els.enemyAbilities.appendChild(li);
            });
        }
    }

    updatePlayerAbilitiesList() {
        this.els.playerAbilities.innerHTML = '';
        if (!this.currentPlayer || !this.currentPlayer.abilities) return;

        this.currentPlayer.abilities.forEach((ability, i) => {
            const li = document.createElement('li');
            const onCooldown = ability.cooldownRemaining > 0;
            li.textContent = onCooldown
                ? `${ability.name} (CD: ${ability.cooldownRemaining})`
                : ability.name;
            if (onCooldown) li.classList.add('on-cooldown');
            if (i === this.selectedAbilityIndex) li.classList.add('selected');

            li.addEventListener('click', () => {
                this.selectedAbilityIndex = i;
                this.updatePlayerAbilitiesList(); // re-render to show selection
            });

            this.els.playerAbilities.appendChild(li);
        });
    }

    showCombatUI() {
        this.els.pnlMovement.classList.add('d-none');
        this.els.pnlCombat.classList.remove('d-none');
        this.els.enemyPanel.classList.remove('d-none');
    }

    showMovementUI() {
        this.els.pnlMovement.classList.remove('d-none');
        this.els.pnlCombat.classList.add('d-none');
        this.els.enemyPanel.classList.add('d-none');

        // Clear enemy display
        this.els.enemyName.textContent = '—';
        this.els.enemyMaxHP.textContent = '—';
        this.els.enemyStr.textContent = 'STR: —';
        this.els.enemyDex.textContent = 'DEX: —';
        this.els.enemyDef.textContent = 'DEF: —';
        this.els.enemyWis.textContent = 'WIS: —';
        this.els.enemyAbilities.innerHTML = '';
    }

    returnToExplorationMode() {
        this.hideActionButton();
        this.showMovementUI();
        this.setMovementEnabled(true);
        this.refreshActionsTakenList();
        this.loadRoom();
    }

    showActionButton(text) {
        this.els.btnAction.textContent = text;
        this.els.btnAction.classList.remove('d-none');
    }

    hideActionButton() {
        this.els.btnAction.classList.add('d-none');
    }

    hasWeaponInInventory() {
        return this.gameState.inventory.includes('Phaser');
    }

    refreshInventoryList() {
        this.els.inventory.innerHTML = '';
        this.gameState.inventory.forEach(item => {
            const li = document.createElement('li');
            li.textContent = item;
            this.els.inventory.appendChild(li);
        });
    }

    refreshActionsTakenList() {
        this.els.actionsTaken.innerHTML = '';
        this.gameState.actionsTaken.forEach(action => {
            const li = document.createElement('li');
            li.textContent = action;
            this.els.actionsTaken.appendChild(li);
        });
    }

    appendLog(message, type = '') {
        const div = document.createElement('div');
        div.classList.add('log-entry');
        if (type === 'heading') div.classList.add('log-heading');
        div.textContent = message;
        this.els.gameLog.appendChild(div);
        this.els.gameLog.scrollTop = this.els.gameLog.scrollHeight;
    }

    // ============ Game State Management ============

    saveGame() {
        this.gameState.playerX = this.playerX;
        this.gameState.playerY = this.playerY;
        this.gameState.playerData = this.currentPlayer.toSaveData();
        this.gameState.save();
    }

    gameOver() {
        if (this.isGameOver) return;
        this.isGameOver = true;

        document.getElementById('modalTitle').textContent = 'Defeat';
        document.getElementById('modalBody').textContent = 'Your mission has failed. Try again?';
        const modal = new bootstrap.Modal(document.getElementById('gameModal'));
        modal.show();
    }

    gameWin() {
        if (this.isGameOver) return;
        this.isGameOver = true;

        this.appendLog('You open the locker and find the shimmering Time Crystal! 💎');
        this.appendLog('With this artifact, you can restore the timeline and save the ship!');
        this.appendLog('MISSION ACCOMPLISHED!');

        document.getElementById('modalTitle').textContent = '🎉 Victory!';
        document.getElementById('modalBody').textContent =
            'You found the Time Crystal and saved the timeline! Play again?';
        const modal = new bootstrap.Modal(document.getElementById('gameModal'));
        modal.show();
    }

    resetGame() {
        GameState.clear();
        switchScreen('screen-select');
    }
}

File: FinalProject/Forms/GameForm.cs
// Final Project for CPT-230-W48 C# Programming I
// Jonathan Coblentz

// Classes: Hero, Villain, Character, Theme Manager (for Star Trek aesthetic)
// Subclasses:
//      Players: choose between Spock, Benjamin Sisko, Worf and Jean Luc-Picard, each with their own abilities and flavor dialogue
//      Enemies:  Bat - A basic easy fight
//                RomulanScout - Another easy fight, and drops a Phaser
//                BorgDrone (not implimented),
//                Tribbles - Tribbles have a 'Cute' defense for the first 4 rounds, they will "multiply" which means their strength
//                and hit points increase exponentially.  They can easily overwhelm the player, but can be beamed out by Scotty
//                Q - performs a few random attacks, teleports Player to another room in the game; if player hasn't fought the Tribbles yet,
//                he sends you there first.  Each player has a unique ability that chases Q away and Q has a snarky response unique to each player.                
// Interfaces:    IAbility, ICombatContext, ICombatant (built from IAttackable and IAttackableTarget)
// Interactive elements:
//      Override console (unlocks a door)
//      Route Power to Sensors (boosts defense)
//      Phaser - boosts damage and blasts through a blocked door
//      Combadge - Calls Scotty.  After two unsuccessful attempts (with a Scotty catchphrase!),
//      it beams the Tribbles into space ending teh combat sequence.
//      KeyCard - opens the captain's locker so the player can take the Time Crystal, ending the gameplay.
// Extra:
//      - Inventory and "Actions taken" log to track the player's progress in the game.  Both are tracked in the GameState object
// which writes to a JSON file so the Player can resume progress if they quit the game.
//      - Player Abilities - They utilize the IAbility interface:
//          Picard: Inspiring Speech, Precision Strike
//          Word: Bat'leth Slash, Warrior's Roar (Worf),
//          Spock: Vulcan Nerve Pinch, Logical Counterattack
//          Sisko: Sucker Punch, Commander's Gambit (Sisko)
//


using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Threading.Tasks;
using System.Windows.Forms;
using FinalProject.Classes;
using FinalProject.GameLogic;
using FinalProject.Interfaces;

namespace FinalProject
{
    public partial class GameForm : Form
    {
        #region Fields & Properties

        public Player CurrentPlayer { get; set; }

        public GameState gameState; // Save data

        private Maze maze; // Build the maze

        private int playerX; // Player coordinates
        private int playerY;

        public CombatEngine CurrentCombat { get; set; }

        private Character CurrentEnemy => CurrentCombat?.CurrentEnemy;

        private bool isGameOver = false;

        public bool ShouldSerializeCurrentPlayer() => false;
        public bool ShouldSerializeCurrentCombat() => false;

        #endregion

        #region Constructors

        // --- Default constructor (for new player) ---
        public GameForm(Player chosenPlayer)
        {
            InitializeComponent();
            ThemeManager.ApplyTrekTheme(this);

            // assign selected player
            CurrentPlayer = chosenPlayer ?? throw new ArgumentNullException(nameof(chosenPlayer));

            // new game state
            gameState = new GameState
            {
                CurrentPlayer = CurrentPlayer,
                PlayerX = playerX,
                PlayerY = playerY
            };

            maze = new Maze();

            // Start with enemy UI hidden
            ShowMovementUI();

            // Populate UI
            UpdatePlayerUI();
            RefreshInventoryList();
            RefreshActionsTakenList();
            EnterRoom();
        }

        // --- Constructor for resuming a saved game ---
        public GameForm(GameState loadedState)
        {
            InitializeComponent();
            ThemeManager.ApplyTrekTheme(this);

            gameState = loadedState ?? new GameState();
            CurrentPlayer = gameState.CurrentPlayer ?? CurrentPlayer;
            playerX = gameState.PlayerX;
            playerY = gameState.PlayerY;

            maze = new Maze();

            // Start with enemy UI hidden until combat starts
            ShowMovementUI();

            UpdatePlayerUI();
            RefreshInventoryList();
            RefreshActionsTakenList();
            EnterRoom();
        }

        #endregion

        #region Core Logic

        public void StartCombat(List<Enemy> enemies)
        {
            CurrentCombat = new CombatEngine(this, CurrentPlayer, enemies);
            CurrentCombat.StartCombat();

            ShowCombatUI();
            btnFight.Visible = true;
            btnFight.Enabled = true;
        }

        private void EnterRoom()
        {
            UpdateMovementButtons();

            txtX.Text = playerX.ToString();
            txtY.Text = playerY.ToString();

            var room = maze.GetRoom(playerX, playerY);
            if (room != null)
            {
                AppendLog($"\r\n=== {room.Name} ===");
                AppendLog(room.Description);

                // Room logic may show/hide action button or do other things
                room.RoomEnterLogic?.Invoke(this);
                room.RoomUILogic?.Invoke(this);
            }
        }

        public void LoadRoom()
        {
            UpdateMovementButtons();

            txtX.Text = playerX.ToString();
            txtY.Text = playerY.ToString();

            maze.GetRoom(playerX, playerY)?.RoomUILogic?.Invoke(this);
        }

        public void TeleportTo(int x, int y)
        {
            playerX = x;
            playerY = y;
            gameState.PlayerX = x;
            gameState.PlayerY = y;
            EnterRoom();
        }

        #endregion

        #region Movement Logic

        private void MovePlayer(int dx, int dy)
        {
            int newX = playerX + dx;
            int newY = playerY + dy;

            var room = maze.GetRoom(newX, newY);
            if (room == null)
            {
                AppendLog("You can't go that way.");
                return;
            }

            playerX = newX;
            playerY = newY;

            if (gameState == null) gameState = new GameState();

            gameState.PlayerX = playerX;
            gameState.PlayerY = playerY;
            gameState.CurrentPlayer = CurrentPlayer;
            gameState.Save("gamestate.json");

            EnterRoom();
        }

        public void SetMovementEnabled(bool enabled)
        {
            if (enabled)
            {
                UpdateMovementButtons();
            }
            else
            {
                btnUp.Enabled = false;
                btnDown.Enabled = false;
                btnLeft.Enabled = false;
                btnRight.Enabled = false;
            }
        }

        public void UpdateMovementButtons()
        {
            btnUp.Enabled = maze.GetRoom(playerX, playerY + 1) != null;
            btnDown.Enabled = maze.GetRoom(playerX, playerY - 1) != null;
            btnRight.Enabled = maze.GetRoom(playerX + 1, playerY) != null;
            btnLeft.Enabled = maze.GetRoom(playerX - 1, playerY) != null;
        }

        public void EnableMovementButtons()
        {
            UpdateMovementButtons();
        }

        public void LockDirection(string direction)
        {
            switch (direction)
            {
                case "Left": btnLeft.Enabled = false; break;
                case "Right": btnRight.Enabled = false; break;
                case "Up": btnUp.Enabled = false; break;
                case "Down": btnDown.Enabled = false; break;
            }
        }

        public void UnlockDirection(string direction)
        {
            switch (direction)
            {
                case "Left": btnLeft.Enabled = true; break;
                case "Right": btnRight.Enabled = true; break;
                case "Up": btnUp.Enabled = true; break;
                case "Down": btnDown.Enabled = true; break;
            }
        }

        #endregion

        #region UI Management

        public void UpdatePlayerUI()
        {
            if (CurrentPlayer == null) return;

            // Ensure progress bar max is set before assigning value
            pbPlayerHealth.Maximum = Math.Max(1, CurrentPlayer.MaxHP);
            pbPlayerHealth.Value = Math.Max(0, Math.Min(CurrentPlayer.CurrentHP, pbPlayerHealth.Maximum));

            lblPlayerName.Text = CurrentPlayer.Name ?? "Player";
            lblPlayerMaxHP.Text = $"{CurrentPlayer.CurrentHP}/{CurrentPlayer.MaxHP}";
            lblPlayerStrength.Text = $"Strength: {CurrentPlayer.Strength}";
            lblPlayerDexterity.Text = $"Dexterity: {CurrentPlayer.Dexterity}";
            lblPlayerDefense.Text = $"Defense: {CurrentPlayer.Defense}";
            lblPlayerWisdom.Text = $"Wisdom: {CurrentPlayer.Wisdom}";

            UpdatePlayerAbilitiesList();
        }

        public void UpdateEnemyUI(Character enemy)
        {
            if (enemy == null)
            {
                ShowMovementUI();
                return;
            }

            ShowCombatUI();
            pbEnemyHealth.Maximum = Math.Max(1, enemy.MaxHP);
            pbEnemyHealth.Value = Math.Max(0, Math.Min(enemy.CurrentHP, pbEnemyHealth.Maximum));

            lblEnemyName.Text = enemy.Name ?? "Enemy";
            lblEnemyMaxHP.Text = $"{enemy.CurrentHP}/{enemy.MaxHP}";
            lblEnemyStrength.Text = $"Strength: {enemy.Strength}";
            lblEnemyDexterity.Text = $"Dexterity: {enemy.Dexterity}";
            lblEnemyDefense.Text = $"Defense: {enemy.Defense}";
            lblEnemyWisdom.Text = $"Wisdom: {enemy.Wisdom}";

            lbEnemySpecialAbilities.Items.Clear();
            if (enemy.Abilities != null)
            {
                lbEnemySpecialAbilities.DisplayMember = "Name";
                foreach (var a in enemy.Abilities)
                    lbEnemySpecialAbilities.Items.Add(a);
            }
        }

        public void ShowCombatUI()
        {
            pnlMovement.Visible = false;
            pnlCombat.Visible = true;

            // Show Enemy Stats
            pnlEnemyStats.Visible = true;
            pbEnemyHealth.Visible = true;
            pBoxEnemy.Visible = true;
        }

        public void ShowMovementUI()
        {
            pnlMovement.Visible = true;
            pnlCombat.Visible = false;

            // Hide Enemy Stats
            pnlEnemyStats.Visible = false;
            pbEnemyHealth.Visible = false;
            pBoxEnemy.Visible = false;

            // Clear enemy display content
            lblEnemyName.Text = "EnemyName";
            lblEnemyMaxHP.Text = "Max HP: ";
            lblEnemyStrength.Text = "Strength: ";
            lblEnemyDexterity.Text = "Dexterity: ";
            lblEnemyDefense.Text = "Defense: ";
            lblEnemyWisdom.Text = "Wisdom: ";
            lbEnemySpecialAbilities.Items.Clear();
        }

        public void ReturnToExplorationMode()
        {
            HideActionButton();
            ShowMovementUI();
            SetMovementEnabled(true);
            RefreshActionsTakenList();
            LoadRoom();
        }

        public bool HasWeaponInInventory()
        {
            return gameState != null && (gameState.Inventory.Contains("Phaser"));
        }

        private void RefreshInventoryList()
        {
            lbInventory.Items.Clear();
            if (gameState == null) return;
            foreach (var item in gameState.Inventory)
                lbInventory.Items.Add(item);
        }

        private void RefreshActionsTakenList()
        {
            lbActionsTaken.Items.Clear();
            if (gameState == null) return;
            foreach (var action in gameState.ActionsTaken)
                lbActionsTaken.Items.Add(action);
        }

        public void AppendLog(string message)
        {
            if (this.IsDisposed || txtGameLog.IsDisposed) return;
            txtGameLog.AppendText("\r\n" + message + "\n\n");
        }

        public void ShowActionButton(string text)
        {
            btnAction.Visible = true;
            btnAction.Text = text;
        }

        public void HideActionButton()
        {
            btnAction.Visible = false;
        }

        public void UpdatePlayerAbilitiesList()
        {
            lbPlayerSpecialAbilities.Items.Clear();
            if (CurrentPlayer?.Abilities == null) return;

            foreach (var ability in CurrentPlayer.Abilities)
            {
                string text = ability.CooldownRemaining > 0
                    ? $"{ability.Name} (CD: {ability.CooldownRemaining})"
                    : ability.Name;

                lbPlayerSpecialAbilities.Items.Add(ability);
                lbPlayerSpecialAbilities.DisplayMember = "Name";
            }
        }

        #endregion

        #region Event Handlers

        private async void btnFight_Click(object sender, EventArgs e)
        {
            if (CurrentCombat == null) return;
            await CurrentCombat.PlayerAttack();
        }

        private async void btnUseAbility_Click(object sender, EventArgs e)
        {
            if (lbPlayerSpecialAbilities.SelectedItem is IAbility ability)
            {
                if (CurrentCombat == null) return;
                await CurrentCombat.PlayerUseAbility(ability);
            }
        }

        protected void btnUp_Click(object sender, EventArgs e) => MovePlayer(0, 1);
        protected void btnDown_Click(object sender, EventArgs e) => MovePlayer(0, -1);
        protected void btnLeft_Click(object sender, EventArgs e) => MovePlayer(-1, 0);
        protected void btnRight_Click(object sender, EventArgs e) => MovePlayer(1, 0);

        private void btnAction_Click(object sender, EventArgs e)
        {
            var room = maze.GetRoom(playerX, playerY);
            if (room == null) return;

            string actionText = btnAction.Text;

            switch (actionText)
            {
                case "Route Power to Sensors":
                    if (!gameState.ActionsTaken.Contains("RoutedPower"))
                        gameState.ActionsTaken.Add("RoutedPower");
                    AppendLog("You reroute auxiliary power to the sensor array. The console lights up.");
                    break;

                case "Take Phaser":
                    if (!gameState.Inventory.Contains("Phaser"))
                        gameState.Inventory.Add("Phaser");
                    AppendLog("You take the Type-2 Phaser.");
                    break;

                case "Cut Blast Door":
                    if (!gameState.ActionsTaken.Contains("CutDoor"))
                        gameState.ActionsTaken.Add("CutDoor");
                    AppendLog("You set the Phaser to maximum and slice through the blast door locks. 💥");
                    AppendLog("The door slides open with a groan.");
                    break;

                case "Engage Override":
                    if (!gameState.ActionsTaken.Contains("Engaged override"))
                    {
                        gameState.ActionsTaken.Add("Engaged override");
                        AppendLog("You engage the manual override. The forcefields deactivate.");
                    }
                    else AppendLog("Override already engaged.");
                    break;

                case "Take Keycard":
                    if (!gameState.Inventory.Contains("Keycard"))
                        gameState.Inventory.Add("Keycard");
                    AppendLog("You pick up the Security Keycard.");
                    break;

                case "Take Tricorder":
                    if (!gameState.Inventory.Contains("Tricorder"))
                        gameState.Inventory.Add("Tricorder");
                    AppendLog("You pick up the standard issue Tricorder.");
                    break;

                case "Access Locker":
                    GameWin();
                    break;

                case "Scan Entity":
                    AppendLog("You scan the entity. It appears to be a silicon-based lifeform.");
                    break;

                case "Take Combadge":
                    if (!gameState.Inventory.Contains("Combadge"))
                    {
                        gameState.Inventory.Add("Combadge");
                        AppendLog("You take the Combadge from the dead Redshirt.");
                        AppendLog("You feel a strange connection to the ship's engineer...");

                        // Grant the ability
                        CurrentPlayer.Abilities.Add(new FinalProject.Players.CallScottyAbility());
                        AppendLog("Ability Acquired: Call Scotty! 🖖");

                        HideActionButton();
                        UpdatePlayerUI();
                    }
                    break;

                default:
                    AppendLog($"Action '{actionText}' not handled.");
                    break;
            }

            // Save and refresh UI
            gameState.Save("gamestate.json");
            LoadRoom();
            RefreshInventoryList();
            RefreshActionsTakenList();
        }

        private void btnStartOver_Click(object sender, EventArgs e)
        {
            var result = MessageBox.Show(
                "Are you sure you want to start over? This will erase all game progress.",
                "Confirm Start Over",
                MessageBoxButtons.YesNo,
                MessageBoxIcon.Warning
            );

            if (result == DialogResult.Yes)
                ResetGame();
        }

        private void btnSaveAndQuit_Click(object sender, EventArgs e)
        {
            gameState.Save("gamestate.json");
            Application.Exit();
        }

        #endregion

        #region Game State Management

        private void ResetGame()
        {
            // Reset in-memory state
            gameState = new GameState();

            // Persist empty/new state
            gameState.Save("gamestate.json");

            // Close current game form (the caller will open CharacterSelectForm)
            this.Close();

            // Open the Character Select screen
            var selectForm = new CharacterSelectForm();
            selectForm.Show();
        }

        public void gameOver()
        {
            if (isGameOver) return;
            isGameOver = true;

            var result = MessageBox.Show(
                "Try again",
                "Try again?",
                MessageBoxButtons.YesNo
                );

            if (result == DialogResult.Yes)
            {
                ResetGame();
            }
            else
            {
                this.Close();
            }
        }

        public void GameWin()
        {
            if (isGameOver) return;
            isGameOver = true;

            MessageBox.Show(
                "You open the locker and find the shimmering Time Crystal! 💎\r\n" +
                "With this artifact, you can restore the timeline and save the ship!\r\n\r\n" +
                "MISSION ACCOMPLISHED!",
                "Victory!",
                MessageBoxButtons.OK,
                MessageBoxIcon.Information
            );

            var result = MessageBox.Show(
                "Play again?",
                "Victory!",
                MessageBoxButtons.YesNo
            );

            if (result == DialogResult.Yes)
            {
                ResetGame();
            }
            else
            {
                this.Close();
            }
        }

        #endregion


    }
}
